utils/bitfield/src/rleplus/reader.rs

```diff
@@ -0,0 +1,119 @@
+    /// The next bits to be read.
+    bits: u16,
+    /// The number of bits in `bits` from bytes that came before `next_byte` (at least 8, at most 15).
+    num_bits: u32,
+}
+
+impl<'a> BitReader<'a> {
+    /// Creates a new `BitReader`.
+    pub fn new(bytes: &'a [u8]) -> Self {
+        let &byte1 = bytes.get(0).unwrap_or(&0);
+        let &byte2 = bytes.get(1).unwrap_or(&0);
+        let bytes = if bytes.len() > 2 { &bytes[2..] } else { &[] };
+
+        Self {
+            bytes,
+            bits: byte1 as u16,
+            next_byte: byte2,
+            num_bits: 8,
+        }
+    }
+
+    /// Reads a given number of bits from the buffer. Will keep returning 0 once
+    /// the buffer has been exhausted.
+    pub fn read(&mut self, num_bits: u32) -> u8 {
+        debug_assert!(num_bits <= 8);
+
+        // creates a mask with a `num_bits` number of 1s in order
+        // to get only the bits we need from `self.bits`
+        let mask = (1 << num_bits) - 1;
+        let res = (self.bits & mask) as u8;
+
+        // removes the bits we've just read from local storage
+        // because we don't need them anymore
+        self.bits >>= num_bits;
+        self.num_bits -= num_bits;
+
+        // this unconditionally adds the next byte to `bits`,
+        // regardless of whether there's enough space or not. the
+        // point is to make sure that `bits` always contains
+        // at least the next 8 bits to be read
+        self.bits |= (self.next_byte as u16) << self.num_bits;
+
+        // if fewer than 8 bits remain, we increment `self.num_bits`
+        // to include the bits from `next_byte` (which is already
+        // contained in `bits`) and we update `next_byte` with the
+        // data to be read after that
+        if self.num_bits < 8 {
+            self.num_bits += 8;
+
+            let (&next_byte, bytes) = self.bytes.split_first().unwrap_or((&0, &[]));
+            self.next_byte = next_byte;
+            self.bytes = bytes;
+        }
+
+        res
+    }
+
+    /// Reads a varint from the buffer. Returns an error if the
+    /// current position on the buffer contains no valid varint.
+    fn read_varint(&mut self) -> Result<usize, &'static str> {
+        let mut len = 0;
+
+        for i in 0..VARINT_MAX_BYTES {
+            let byte = self.read(8);
+
+            // strip off the most significant bit and add
+            // it to the output
+            len |= (byte as usize & 0x7f) << (i * 7);
+
+            // if the most significant bit is a 0, we've
+            // reached the end of the varint
+            if byte & 0x80 == 0 {
+                return Ok(len);
+            }
+        }
+
+        Err("Invalid varint")
+    }
+
+    /// Reads a length from the buffer according to RLE+ encoding.
+    pub fn read_len(&mut self) -> Result<Option<usize>, &'static str> {
+        let len = if self.read(1) == 1 {
+            // Block Single (prefix 1)
+            Some(1)
+        } else if self.read(1) == 1 {
```

Comment:
You if and else if conditional is the same

Reply:
Yeah it looks a little weird, but the second `read` reads the bit after that.
The `if`/`else if`/`else` here reads the prefix which is either `1`, `01`, or `00`

Reply:
I'll put them in separate variables, that will make it clearer

Reply:
I figured that the intention based on the comments, but i still dont see how it will EVER reach the else if since the conditional is the same as the if

Reply:
`self.read` advances `self`, so calling `self.read(1)` twice in a row won't read the same bit, it will read different bits — and the `self.read(1)` in the `else` will obviously only be called after the first `self.read(1)` has returned `0`

Reply:
Yeah, I get that. Ok, maybe I'm totally not understanding something here...

Reply:
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOK Wow, i see whats happening